File: Trees/0129. Sum Root to Leaf Numbers/0129.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        def dfs(node, currNum):
            if not node.left and not node.right:
                return node.val + currNum
            res=0
            if node.left:
                res+=dfs(node.left, (node.val+currNum)*10)
            if node.right:
                res+=dfs(node.right, (node.val+currNum)*10)
            return res
            
        return dfs(root,0)
            

File: Trees/0129. Sum Root to Leaf Numbers/0129.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
#pragma GCC optimize('03','unroll-loops') 

class Solution {
public:
    int dfs(TreeNode* node, int currNum) {
        if(!node->left && !node->right){
            return node->val + currNum;
        }
        int res=0;
        if(node->left){
            res+=dfs(node->left, (node->val+currNum)*10);
        }
        if(node->right){
            res+=dfs(node->right, (node->val+currNum)*10);
        }
        return res;
    }

    int sumNumbers(TreeNode* root) {
        ios_base::sync_with_stdio(0);
        cin.tie(0);
        cout.tie(0);

        return dfs(root, 0);
    }
};